From: Nathan Pratta Teodosio <nathan.teodosio@canonical.com>
Date: Thu, 29 Aug 2024 09:25:40 +0200
Subject: Use 'access' instead of checking permission modes for
 /tmp/.X11-unix/.

The previous version was missing the 200 case with the directory owned by the
current user.

C.f. https://gitlab.gnome.org/GNOME/gnome-shell/-/issues/7857,
     https://launchpad.net/bugs/2069564.
---
 src/wayland/meta-xwayland.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

diff --git a/src/wayland/meta-xwayland.c b/src/wayland/meta-xwayland.c
index 4a0fe4f..681b2ad 100644
--- a/src/wayland/meta-xwayland.c
+++ b/src/wayland/meta-xwayland.c
@@ -545,14 +545,22 @@ ensure_x11_unix_perms (GError **error)
   if (x11_tmp.st_uid != tmp.st_uid && x11_tmp.st_uid != getuid ())
     {
       g_set_error (error, G_IO_ERROR, G_IO_ERROR_PERMISSION_DENIED,
-                   "Wrong ownership for directory \"%s\"",
-                   X11_TMP_UNIX_DIR);
+                   "Wrong ownership for directory \"%s\": %d.\n"
+                   "Ownership of \"%s\": %d.\n"
+                   "Current user: %d.",
+                   X11_TMP_UNIX_DIR, x11_tmp.st_uid, TMP_UNIX_DIR, tmp.st_uid,
+                   getuid ());
       return FALSE;
     }
 
   /* ... be writable ... */
-  if ((x11_tmp.st_mode & 0022) != 0022)
+  if (access (X11_TMP_UNIX_DIR, W_OK) != 0)
     {
+      if ((x11_tmp.st_mode & 0022) != 0022)
+        {
+          g_warning ("%s is accessible although its permissions are not 0022",
+                     X11_TMP_UNIX_DIR);
+        }
       g_set_error (error, G_IO_ERROR, G_IO_ERROR_PERMISSION_DENIED,
                    "Directory \"%s\" is not writable",
                    X11_TMP_UNIX_DIR);
